### PR TITLE
[tools/onert_train] Revise args to hold NNFW_TRAIN_* enum

### PR DESCRIPTION
**Pull request by zetwhite**
This PR revises Args class to hold NNFW_TRAIN_* enum directly, instead of holding integer value.

ONE-DCO-1.0-Signed-off-by: SeungHui Youn <sseung.youn@samsung.com>

draft : https://github.com/Samsung/ONE/pull/12445 
in the process to resolve : https://github.com/Samsung/ONE/issues/12520

### CODE REVIEW

File: tests/tools/onert_train/src/args.h
Comment (zetwhite) [inline tests/tools/onert_train/src/args.h:89]: This PR"s main diff is here : use `NNFW_TRAIN_*` instead of integer value.

File: tests/tools/onert_train/src/nnfw_util.h
Comment (zetwhite) [inline tests/tools/onert_train/src/nnfw_util.h:49]: (note) to find the underlying value(usually integer) for enum

(I referred this function from modern c++ item 10)

File: tests/tools/onert_train/src/args.cc
Comment (zetwhite) [inline tests/tools/onert_train/src/args.cc:209]: If `loss` is introduced in training and to support new loss in `tools/onert_train`, we can do it by adding one entity here.

File: tests/tools/onert_train/src/args.cc
Comment (zetwhite) [inline tests/tools/onert_train/src/args.cc:309]: (note) 

After this, onert_train help messages slightly changed. 

```diff
➜ diff master_br_help current_br_help
<   --loss arg (=1)                       Loss type
<                                         1: MEAN_SQUARED_ERROR (default)
<                                         2: CATEGORICAL_CROSSENTROPY
<   --loss_reduction_type arg (=1)        Loss Reduction type
<                                         1: SUM_OVER_BATCH_SIZE(default)
<                                         2: SUM
<   --optimizer arg (=1)                  Optimizer type
<                                         1: SGD (default)
<                                         2: Adam
---
>   --loss arg (=1)                       Loss type (default: mean squared error)
>                                         1: mean squared error
>                                         2: categorical crossentropy
>   --loss_reduction_type arg (=1)        Loss Reduction type (default: sum over batch size)
>                                         1: sum over batch size
>                                         2: sum
>   --optimizer arg (=1)                  Optimizer (default: sgd)
>                                         1: sgd
>                                         2: adam
```


<details> 

* master branch help message 
	```
	onert_train
	
	Usage: /home/sseung/Workspace/ONE/master/Product/build/bin/onert_train [model path] [<options>]
	
	
	General options:
	  -h [ --help ]                         Print available options
	  --version                             Print version and exit immediately
	  --nnpackage arg                       NN Package file(directory) name
	  --modelfile arg                       NN Model filename
	  --path arg                            NN Package or NN Modelfile path
	  --export_path arg                     Path to export circle
	  --load_input:raw arg                  NN Model Raw Input data file
	                                        The datafile must have data for each input number.
	                                        If there are 3 inputs, the data of input0 must exist as 
	                                        much as data_length, and the data for input1 and input2 
	                                        must be held sequentially as data_length.
	  --load_expected:raw arg               NN Model Raw Expected data file
	                                        (Same data policy with load_input:raw)
	  -m [ --mem_poll ] arg (=0)            Check memory polling (default: false)
	  --epoch arg (=5)                      Epoch number (default: 5)
	  --batch_size arg (=32)                Batch size (default: 32)
	  --learning_rate arg (=0.00100000005)  Learning rate (default: 0.001)
	  --loss arg (=1)                       Loss type
	                                        1: MEAN_SQUARED_ERROR (default)
	                                        2: CATEGORICAL_CROSSENTROPY
	  --loss_reduction_type arg (=1)        Loss Reduction type
	                                        1: SUM_OVER_BATCH_SIZE(default)
	                                        2: SUM
	  --optimizer arg (=1)                  Optimizer type
	                                        1: SGD (default)
	                                        2: Adam
	  --metric arg (=-1)                    Metric type
	                                        Simply calculates the metric value using the variables 
	                                        (default: none)
	                                        0: CATEGORICAL_ACCURACY
	  --validation_split arg (=0)           Float between 0 and 1(0 < float < 1). Fraction of the 
	                                        training data to be used as validation data.
	  -v [ --verbose_level ] arg (=0)       Verbose level
	                                        0: prints the only result. Messages btw run don't print
	                                        1: prints result and message btw run
	                                        2: prints all of messages to print
	  --output_sizes arg                    The output buffer size in JSON 1D array
	                                        If not given, the model's output sizes are used
	                                        e.g. '[0, 40, 2, 80]' to set 0th tensor to 40 and 2nd 
	                                        tensor to 80.
	
	```

* current branch help message
	```
	onert_train
	
	Usage: /home/sseung/Workspace/ONE/current/Product/build/bin/onert_train [model path] [<options>]
	
	
	General options:
	  -h [ --help ]                         Print available options
	  --version                             Print version and exit immediately
	  --nnpackage arg                       NN Package file(directory) name
	  --modelfile arg                       NN Model filename
	  --path arg                            NN Package or NN Modelfile path
	  --export_path arg                     Path to export circle
	  --load_input:raw arg                  NN Model Raw Input data file
	                                        The datafile must have data for each input number.
	                                        If there are 3 inputs, the data of input0 must exist as 
	                                        much as data_length, and the data for input1 and input2 
	                                        must be held sequentially as data_length.
	  --load_expected:raw arg               NN Model Raw Expected data file
	                                        (Same data policy with load_input:raw)
	  -m [ --mem_poll ] arg (=0)            Check memory polling (default: false)
	  --epoch arg (=5)                      Epoch number (default: 5)
	  --batch_size arg (=32)                Batch size (default: 32)
	  --learning_rate arg (=0.00100000005)  Learning rate (default: 0.001)
	  --loss arg (=1)                       Loss type (default: mean squared error)
	                                        1: mean squared error
	                                        2: categorical crossentropy
	  --loss_reduction_type arg (=1)        Loss Reduction type (default: sum over batch size)
	                                        1: sum over batch size
	                                        2: sum
	  --optimizer arg (=1)                  Optimizer (default: sgd)
	                                        1: sgd
	                                        2: adam
	  --metric arg (=-1)                    Metric type
	                                        Simply calculates the metric value using the variables 
	                                        (default: none)
	                                        0: CATEGORICAL_ACCURACY
	  --validation_split arg (=0)           Float between 0 and 1(0 < float < 1). Fraction of the 
	                                        training data to be used as validation data.
	  -v [ --verbose_level ] arg (=0)       Verbose level
	                                        0: prints the only result. Messages btw run don't print
	                                        1: prints result and message btw run
	                                        2: prints all of messages to print
	  --output_sizes arg                    The output buffer size in JSON 1D array
	                                        If not given, the model's output sizes are used
	                                        e.g. '[0, 40, 2, 80]' to set 0th tensor to 40 and 2nd 
	                                        tensor to 80.
	```
</details>

**Comment by zetwhite** [comment]
@jyoungyun, @YongseopKim 

Could you take a look? Thank you in advance. :smile: 
I'd like to get your approvals(ofc, other opinions are also welcome!) - because you're the one who recently modified `onert_train/args`.

File: tests/tools/onert_train/src/onert_train.cc
Comment (zetwhite) [inline tests/tools/onert_train/src/onert_train.cc:162]: This PR's motivation is here: Remove this hard-coded parsing in onert_train's main function.

File: tests/tools/onert_train/src/nnfw_util.cc
Comment (jyoungyun) [inline tests/tools/onert_train/src/nnfw_util.cc:119]: You don't need this function here, do you?

Comment (zetwhite) [reply tests/tools/onert_train/src/nnfw_util.cc:119]: Ah, Yess. 
My mistake. I'll remove it!

**Review by jyoungyun — DISMISSED** [review]
LGTM

File: tests/tools/onert_train/src/args.cc
Comment (zetwhite) [inline tests/tools/onert_train/src/args.cc:268]: Ah, I found an error in this part. 
I'm trying to reconfigure ..

File: tests/tools/onert_train/src/nnfw_util.h
Comment (YongseopKim) [reply tests/tools/onert_train/src/nnfw_util.h:49]: Oh! I've never seen this before. Interesting!

Comment (YongseopKim) [reply tests/tools/onert_train/src/nnfw_util.h:49]: Naming is `to_underlying`(snake). Above functions' names are `toString` (camel). Which one is right?

**Review by YongseopKim — COMMENTED** [review]
Hmm.. to me, it seems so big and complex so I don't feel easy when reviewing. How about separating PR?
1. introduce helper function `to_underlying` with reasons
2. Add/revise args code
3. Apply revised args code

If you want to continue with 1 PR, of course, I can review this as it is.

**Comment by zetwhite** [comment]
> Hmm.. to me, it seems so big and complex so I don't feel easy when reviewing. How about separating PR?
 
Aha, Thank you for feedback.
I'd better to retry with smaller PR!

File: tests/tools/onert_train/src/args.cc
Comment (zetwhite) [inline tests/tools/onert_train/src/args.cc:113]: check given integer is in a `valid_args` or not.

**Comment by zetwhite** [comment]
@YongseopKim 

Thanks for your offline feedback. :smile: 
I updated this draft a bit. 

I tried to find another way, but for me, it's hard to think otherwise.
So, If you are fine with this draft, I'd like to process as it is.

( But I don't like to push the way someone opposes! So please feel free to give any opinion )

File: tests/tools/onert_train/src/args.cc
Comment (YongseopKim) [inline tests/tools/onert_train/src/args.cc:233]: Can I ask why you declare `static` on it?

Comment (YongseopKim) [reply tests/tools/onert_train/src/args.cc:233]: AFAIK, `const` looks much better.

File: tests/tools/onert_train/src/args.cc
Comment (YongseopKim) [inline tests/tools/onert_train/src/args.cc:280]: It looks little difficult but so much better.

Let's get our goal to achieve simpler like in the future.
```cpp
    ("nnpackage", po::value<std::string>()->notifier(process_nnpackage), "NN Package file(directory) name")
    ("modelfile", po::value<std::string>()->notifier(process_modelfile), "NN Model filename")
```

My idea is making simpler below code
```cpp
// as-is
    ("loss_reduction_type", po::value<int>()
      ->default_value(default_loss_rdt)
      ->notifier([&] (const auto &v) {_loss_reduction_type = check_validation<NNFW_TRAIN_LOSS_REDUCTION>(valid_loss_rdt, v, "loss_reduction_type");}),
      gen_help_msg<NNFW_TRAIN_LOSS_REDUCTION>("loss_reduction_tye", valid_loss_rdt, default_loss_rdt).c_str()
    )

// to-be
    ("loss_reduction_type", po::value<int>()
      ->default_value(default_loss_rdt)
      ->notifier(process_loss_reduction_type),
      help_msg_loss_reduction_type()
    )
```

(I don't think the time is now. Just opinion. 😄)

**Review by YongseopKim — APPROVED** [review]
LGTM 👍 If you're ready, make it to PR.

File: tests/tools/onert_train/src/args.cc
Comment (zetwhite) [reply tests/tools/onert_train/src/args.cc]: I added `const`, here. 

> Can I ask why you declare static on it? 

`std::vector<NNFW_TRAIN_*>` reference is captured by the lambda. 
Since the lambda invoked when `Args::Parse` called, `std::vector<NNFW_TRAIN_*>` have to alive after `Args::Initialized()` is called. 
That's why I added static here.

**Comment by YongseopKim** [comment]
If you believe this is ready for review, transit the state from `draft(work in progress)` to `Ready for review`.

File: tests/tools/onert_train/src/nnfw_util.h
Comment (zetwhite) [reply tests/tools/onert_train/src/nnfw_util.h]: :+1: :+1: 
resolved by : https://github.com/Samsung/ONE/pull/12577

File: tests/tools/onert_train/src/args.cc
Comment (zetwhite) [reply tests/tools/onert_train/src/args.cc]: Hmm. I rethink this part. 
Since the `static` variable is alive throughout the whole program's lifetime. 
Instead of using `static`, I moved it into a class member.

File: tests/tools/onert_train/src/args.cc
Comment (zetwhite) [inline tests/tools/onert_train/src/args.cc:271]: (note) 

After this, onert_train help messages slightly changed. 

```diff
➜ diff master_br_help current_br_help
<   --loss arg (=1)                       Loss type
<                                         1: MEAN_SQUARED_ERROR (default)
<                                         2: CATEGORICAL_CROSSENTROPY
<   --loss_reduction_type arg (=1)        Loss Reduction type
<                                         1: SUM_OVER_BATCH_SIZE(default)
<                                         2: SUM
<   --optimizer arg (=1)                  Optimizer type
<                                         1: SGD (default)
<                                         2: Adam
---
>   --loss arg (=1)                       Loss type (default: mean squared error)
>                                         1: mean squared error
>                                         2: categorical crossentropy
>   --loss_reduction_type arg (=1)        Loss Reduction type (default: sum over batch size)
>                                         1: sum over batch size
>                                         2: sum
>   --optimizer arg (=1)                  Optimizer (default: sgd)
>                                         1: sgd
>                                         2: adam
```


<details> 

* master branch help message 
	```
	onert_train
	
	Usage: /home/sseung/Workspace/ONE/master/Product/build/bin/onert_train [model path] [<options>]
	
	
	General options:
	  -h [ --help ]                         Print available options
	  --version                             Print version and exit immediately
	  --nnpackage arg                       NN Package file(directory) name
	  --modelfile arg                       NN Model filename
	  --path arg                            NN Package or NN Modelfile path
	  --export_path arg                     Path to export circle
	  --load_input:raw arg                  NN Model Raw Input data file
	                                        The datafile must have data for each input number.
	                                        If there are 3 inputs, the data of input0 must exist as 
	                                        much as data_length, and the data for input1 and input2 
	                                        must be held sequentially as data_length.
	  --load_expected:raw arg               NN Model Raw Expected data file
	                                        (Same data policy with load_input:raw)
	  -m [ --mem_poll ] arg (=0)            Check memory polling (default: false)
	  --epoch arg (=5)                      Epoch number (default: 5)
	  --batch_size arg (=32)                Batch size (default: 32)
	  --learning_rate arg (=0.00100000005)  Learning rate (default: 0.001)
	  --loss arg (=1)                       Loss type
	                                        1: MEAN_SQUARED_ERROR (default)
	                                        2: CATEGORICAL_CROSSENTROPY
	  --loss_reduction_type arg (=1)        Loss Reduction type
	                                        1: SUM_OVER_BATCH_SIZE(default)
	                                        2: SUM
	  --optimizer arg (=1)                  Optimizer type
	                                        1: SGD (default)
	                                        2: Adam
	  --metric arg (=-1)                    Metric type
	                                        Simply calculates the metric value using the variables 
	                                        (default: none)
	                                        0: CATEGORICAL_ACCURACY
	  --validation_split arg (=0)           Float between 0 and 1(0 < float < 1). Fraction of the 
	                                        training data to be used as validation data.
	  -v [ --verbose_level ] arg (=0)       Verbose level
	                                        0: prints the only result. Messages btw run don't print
	                                        1: prints result and message btw run
	                                        2: prints all of messages to print
	  --output_sizes arg                    The output buffer size in JSON 1D array
	                                        If not given, the model's output sizes are used
	                                        e.g. '[0, 40, 2, 80]' to set 0th tensor to 40 and 2nd 
	                                        tensor to 80.
	
	```

* current branch help message
	```
	onert_train
	
	Usage: /home/sseung/Workspace/ONE/current/Product/build/bin/onert_train [model path] [<options>]
	
	
	General options:
	  -h [ --help ]                         Print available options
	  --version                             Print version and exit immediately
	  --nnpackage arg                       NN Package file(directory) name
	  --modelfile arg                       NN Model filename
	  --path arg                            NN Package or NN Modelfile path
	  --export_path arg                     Path to export circle
	  --load_input:raw arg                  NN Model Raw Input data file
	                                        The datafile must have data for each input number.
	                                        If there are 3 inputs, the data of input0 must exist as 
	                                        much as data_length, and the data for input1 and input2 
	                                        must be held sequentially as data_length.
	  --load_expected:raw arg               NN Model Raw Expected data file
	                                        (Same data policy with load_input:raw)
	  -m [ --mem_poll ] arg (=0)            Check memory polling (default: false)
	  --epoch arg (=5)                      Epoch number (default: 5)
	  --batch_size arg (=32)                Batch size (default: 32)
	  --learning_rate arg (=0.00100000005)  Learning rate (default: 0.001)
	  --loss arg (=1)                       Loss type (default: mean squared error)
	                                        1: mean squared error
	                                        2: categorical crossentropy
	  --loss_reduction_type arg (=1)        Loss Reduction type (default: sum over batch size)
	                                        1: sum over batch size
	                                        2: sum
	  --optimizer arg (=1)                  Optimizer (default: sgd)
	                                        1: sgd
	                                        2: adam
	  --metric arg (=-1)                    Metric type
	                                        Simply calculates the metric value using the variables 
	                                        (default: none)
	                                        0: CATEGORICAL_ACCURACY
	  --validation_split arg (=0)           Float between 0 and 1(0 < float < 1). Fraction of the 
	                                        training data to be used as validation data.
	  -v [ --verbose_level ] arg (=0)       Verbose level
	                                        0: prints the only result. Messages btw run don't print
	                                        1: prints result and message btw run
	                                        2: prints all of messages to print
	  --output_sizes arg                    The output buffer size in JSON 1D array
	                                        If not given, the model's output sizes are used
	                                        e.g. '[0, 40, 2, 80]' to set 0th tensor to 40 and 2nd 
	                                        tensor to 80.
	```
</details>

Comment (zetwhite) [reply tests/tools/onert_train/src/args.cc:271]: If invalid input is given, it shows below messages. 

```bash 
➜ ~/Workspace/ONE_second/Product/build/bin/onert_train \
--modelfile mnist_with_meta.circle \
--load_input:raw pool_mnist_model/mnist_x_train.1000.bin \
--load_expected:raw pool_mnist_model/mnist_y_train.1000.bin \
--loss_reduction_type 3   
Model Expected Filename pool_mnist_model/mnist_y_train.1000.bin
Model Input Filename pool_mnist_model/mnist_x_train.1000.bin
loss_reduction_type 3 is unsupported argument
```

**Comment by zetwhite** [comment]
> If you believe this is ready for review, transit the state from `draft(work in progress)` to `Ready for review`.

PTAL :smile: 
Thank you in advance. 

@jyoungyun @YongseopKim

**Review by YongseopKim — APPROVED** [review]
LGTM 👍

**Review by jyoungyun — APPROVED** [review]
LGTM!

**Pull request merged**
